### PR TITLE
Context pred GNN Modular

### PR DESCRIPTION
**Pull request by tonydavis629**
## Description

Implements the context prediction task for GNNModular from Strategies for Pretraining Graph Neural Networks by Hu et al 2020. 

Context prediction is the task of predicting the surrounding nodes and edges of a chosen root node. Empirical results show better performance over masking nodes or edges.

## Type of change

Please check the option that is related to your PR.

- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
- [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
- [x] Run `mypy -p deepchem` and check no errors
- [x] Run `flake8 <modified file> --count` and check no errors
- [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/gnn.py [hunk 1]
```diff
@@ -559,7 +622,80 @@ def infomax_loss(self, inputs):
         negative_score = self.components['discriminator'](
             node_emb, negative_expanded_summary_emb)
 
-        return self.graph_infomax_loss(positive_score, negative_score)
+        return self.infomax_loss(positive_score, negative_score)
+
+    def context_pred_loss_loader(self, inputs, labels):
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:627]: Docstring

**Review by rbharath — COMMENTED** [review]
Let's break out the graphdata changes into their own smaller PR for simplicity

File: deepchem/feat/graph_data.py [hunk 2]
```diff
@@ -254,6 +254,52 @@ def numpy_to_torch(self, device: str = 'cpu'):
 
         return graph_copy
 
+    def subgraph(self, nodes):
+        """Returns a subgraph of `nodes` indicies.
+
```
[the remark this answers is not in the document]
Comment (tonydavis629) [reply deepchem/feat/graph_data.py:259]: done

File: deepchem/feat/graph_data.py [hunk 3]
```diff
@@ -408,3 +455,67 @@ def numpy_to_torch(self, device: str = "cpu"):
         graph_copy.graph_index = graph_index
 
         return graph_copy
+
+
+def shortest_path_length(graph_data, source, cutoff=None):
```
[the remark this answers is not in the document]
Comment (tonydavis629) [reply deepchem/feat/graph_data.py:471]: done

File: deepchem/models/losses.py [hunk 4]
```diff
@@ -1168,6 +1168,95 @@ def loss(positive_score, negative_score):
...
+
+                shifted_expanded_substruct_rep = torch.cat(
+                    shifted_expanded_substruct_rep, dim=0)
+                pred_neg = torch.sum(shifted_expanded_substruct_rep *
+                                     overlapped_node_rep.repeat(
+                                         (self.neg_samples, 1)),
+                                     dim=1)
+
+            else:
+                raise ValueError("Invalid mode!")
```
[the remark this answers is not in the document]
Comment (tonydavis629) [reply deepchem/models/losses.py:1245]: done

File: deepchem/models/tests/test_losses.py [hunk 5]
```diff
@@ -546,3 +546,37 @@ def test_deep_graph_infomax_loss(self):
...
+
+        graph_context_pred_loss = GraphContextPredLoss()._create_pytorch_loss(
+            mode, neg_samples)
+
+        loss = graph_context_pred_loss(substruct_rep, overlapped_node_rep,
+                                       context_rep, neg_context_rep,
+                                       overlapped_context_size)
+
+        assert loss.dim() == 0
+        assert loss.dtype == torch.float64
```
[the remark this answers is not in the document]
Comment (tonydavis629) [reply deepchem/models/tests/test_losses.py]: done

File: deepchem/models/losses.py [hunk 6]
```diff
@@ -1168,6 +1168,95 @@ def loss(positive_score, negative_score):
...
+
+            #  Contexts are represented by
+            if self.mode == "cbow":
+                # positive context representation
+                pred_pos = torch.sum(substruct_rep * context_rep, dim=1)
+                pred_neg = torch.sum(substruct_rep.repeat(
+                    (self.neg_samples, 1)) * neg_context_rep,
+                                     dim=1)
+
+            elif self.mode == "skipgram":
```
[the remark this answers is not in the document]
Comment (tonydavis629) [reply deepchem/models/losses.py:1217]: done

File: deepchem/models/torch_models/gnn.py [hunk 7]
```diff
@@ -397,7 +396,7 @@ def build_components(self):
                        self.jump_knowledge)
 
         if self.task in ("mask_nodes", "mask_edges"):
-            self.emb_dim = (self.num_layer + 1) * self.emb_dim
+            # self.emb_dim = (self.num_layer + 1) * self.emb_dim
```
[the remark this answers is not in the document]
Comment (tonydavis629) [reply deepchem/models/torch_models/gnn.py:399]: done

**Review by rbharath — APPROVED** [review]
LGTM
